tests/testthat/test-verb-group_by.R

```diff
@@ -48,9 +48,36 @@ test_that("group_by handles empty dots", {
   lf <- lazy_frame(x = 1) %>% group_by(x)
 
   expect_equal(lf %>% group_by() %>% group_vars(), character())
+  expect_equal(lf %>% group_by(!!!list()) %>% group_vars(), character())
   expect_equal(lf %>% group_by(.add = TRUE) %>% group_vars(), c("x"))
 })
 
+test_that("only adds step if necessary", {
+  lf <- lazy_frame(x = 1, y = 1)
+  expect_equal(lf %>% group_by(), lf)
+
+  expect_equal(lf %>% ungroup(), lf)
+  expect_equal(lf %>% ungroup(x), lf)
+
+  lf_grouped <- lf %>% group_by(x)
+  lf_grouped2 <- lf_grouped %>% group_by(x)
+  expect_equal(lf_grouped, lf_grouped2)
+  expect_equal(lf_grouped %>% ungroup(y), lf_grouped)
+
+  out <- lf_grouped %>% mutate(y = y - mean(y)) %>% group_by()
+  expect_s3_class(out$ops, "op_group_by")
+  expect_equal(group_vars(out), character())
+})
+
+test_that("group_by mutate is not optimised away", {
+  lf <- lazy_frame(x = 1) %>% group_by(x)
+
+  expect_equal(
```

Comment:
Should this be a snapshot test?

Reply:
I don't think so: the expectation is very simple, I know it explicitly, and I don't expect this to change.